uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/layout/dlm/remoting/ChannelListController.java

```diff
@@ -528,4 +573,92 @@ public void setDescription(String descr) {
+
+    private Map<String, SortedSet<PortletCategoryBean>> filterRegistryFavoritesOnly(
+            Map<String, SortedSet<PortletCategoryBean>> registry, HttpServletRequest request) {
+
+        /*
+         * It's not fantastic, but the storage strategy for favorites in the layout XML doc.  We
+         * have to use it to know which portlets are favorites.
+         */
+        final IUserInstance ui = userInstanceManager.getUserInstance(request);
+        final UserPreferencesManager upm = (UserPreferencesManager) ui.getPreferencesManager();
+        final IUserLayoutManager ulm = upm.getUserLayoutManager();
+        final IUserLayout layout = ulm.getUserLayout();
+        final Set<IPortletDefinition> favoritePortlets =
+                favoritesUtils.getFavoritePortletDefinitions(layout);
+
+        logger.debug(
+                "Found the following favoritePortlets for user='{}':  {}",
+                request.getRemoteUser(),
+                favoritePortlets);
+
+        final Set<PortletCategoryBean> inpt = registry.get(CATEGORIES_MAP_KEY);
+        final SortedSet<PortletCategoryBean> otpt = new TreeSet<>();
+        inpt.forEach(
+                categoryIn -> {
+                    final PortletCategoryBean categoryOut =
+                            filterCategoryFavoritesOnly(categoryIn, favoritePortlets);
+                    if (categoryOut != null) {
+                        otpt.add(categoryOut);
+                    }
+                });
+
+        final Map<String, SortedSet<PortletCategoryBean>> rslt = new TreeMap<>();
+        rslt.put(CATEGORIES_MAP_KEY, otpt);
+        return rslt;
+    }
+
+    /**
+     * Returns the filtered category, or <code>null</code> if there is no content remaining in the
+     * category.
+     */
+    private PortletCategoryBean filterCategoryFavoritesOnly(
+            PortletCategoryBean category, Set<IPortletDefinition> favoritePortlets) {
+
+        // Subcategories
+        final Set<PortletCategoryBean> subcategories = new HashSet<>();
+        category.getSubcategories()
+                .forEach(
+                        sub -> {
+                            final PortletCategoryBean filteredBean =
+                                    filterCategoryFavoritesOnly(sub, favoritePortlets);
+                            if (filteredBean != null) {
+                                subcategories.add(filteredBean);
+                            }
+                        });
+
+        // Portlets
+        final Set<PortletDefinitionBean> portlets = new HashSet<>();
+        category.getPortlets()
+                .forEach(
+                        child -> {
+                            if (isPortletAFavorite(child, favoritePortlets)) {
+                                logger.debug(
+                                        "Including portlet '{}' because it is a favorite:  {}",
+                                        child.getFname());
+                                portlets.add(child);
```

Comment:
Could/should this `forEach` be a `filter` and `collect`?
https://www.mkyong.com/java8/java-8-streams-filter-examples/

Reply:
I'm still not terrible fond of the log everything approach over using breakpoints.